add: naive implementation of PPO_LLM

ss26 opened the pull request:
additions:

**ppo llm**
- created `ppo_llm.py` closely following the BaseAgent structure that's currently used in `ppo.py`, using the notebook code from Jon and Nathaniel
- created `test_ppo_llm.py` to test basic functionality of each class in `ppo_llm.py`. all basic tests pass. 
    - to run these tests, check the usage command at the top of `test_ppo_llm.py`. run from the root directory. 
- added code to `train.py` for training using the new PPO LLM agent, HOWEVER i couldn't test the training because it exceeds my laptop's GPU memory :(

possible errors:
- since i couldn't train on my laptop, i'm unsure if the code works completely or if the ppo-llm agent will train at all, this is simply a starting point 

**import refactor**
- minor refactor of imports so running code from the `root_dir` can be centrally done using `python -m <exec command>` 
- updated the readme and added USAGE commands on the top of each relevant file

## Summary by Sourcery

Add a naive PPO_LLM implementation alongside import and CLI refactors to support language-model-based PPO training and evaluation from the project root.

New Features:
- Introduce PPO_LLM agent with ActorNetwork, CriticNetwork, and PPOLLMAgent classes for language-model-based PPO.
- Integrate PPO_LLM into train.py with custom trajectory collection, agent selection, and separate actor/critic checkpointing.

Enhancements:
- Refactor imports to use `src.` module paths and enable running scripts via `python -m` from the project root.

Documentation:
- Update README usage instructions to use `python -m src.<module>` syntax for training and evaluation commands.

Tests:
- Add comprehensive tests for ActorNetwork, CriticNetwork, and PPOLLMAgent covering forward passes, action selection, trajectory collection, GAEs, and update logic.

nmitrani commented on src/agent/ppo_llm.py:
```diff
@@ -0,0 +1,285 @@
+        state_tokens = {k: v.to(device) for k, v in state_tokens.items()}
+        action_tokens = {k: v.to(device) for k, v in action_tokens.items()}
+        action_tokens["input_ids"] = action_tokens["input_ids"][:, 1:]  # Remove BOS
+        action_tokens["input_ids"] = torch.cat(
```
I don't understand why you would pad the tensor with zeros for each action to then set it at 1 (EOS). Isn't it simpler to pad it with torch.ones? Similarly, I don't understand when you add the indexing to the first dim of the action_tokens tensor [torch.arange(batch_size),...] Isn't the first dimension already of size batch_size, and could you not simply do: action_tokens[:,...]?

nmitrani commented on src/agent/ppo_llm.py:
```diff
@@ -0,0 +1,285 @@
+            # Add a linear head if not present
+            if not hasattr(self, 'lm_head'):
+                hidden_size = outputs.last_hidden_state.shape[-1]
+                self.lm_head = nn.Linear(hidden_size, self.tokenizer.vocab_size).to(device)
```
What models do not have a linear head? I had never heard or worked with any such model. If it's the case that they are common, good catch!

nmitrani commented on src/agent/ppo_llm.py:
```diff
@@ -0,0 +1,285 @@
+        logits = torch.clamp(logits, min=-100, max=100)
+        k = state_tokens["input_ids"].shape[1]
+        m = action_tokens["input_ids"].shape[1]
+        next_token_logits = logits[:, k - 1 : -1, :]
```
Before you add an EOS, but here we don't count it towards the calculation right? The issue whether to include it or not is one we talked with Jon, so we can sync on that if you'd like, but for now it seems like you're adding it and then not taking it into account for computations

ss26 replied:
i'll discuss the EOS stuff with you on the call in greater detail. i ported the code from the notebook as-is, and i have some opinions about its place in the action tokens

nmitrani replied:
awesome, let's talk about it yes!

nmitrani commented on src/agent/ppo_llm.py:
```diff
@@ -0,0 +1,285 @@
+        super().__init__()
+        self.model = model
+        self.tokenizer = tokenizer
+        self.simple_nn = nn.Sequential(
+            nn.Flatten(),
+            nn.Linear(input_dim, 64),
+            nn.ReLU(),
+            nn.Linear(64, 64),
+            nn.ReLU(),
+            nn.Linear(64, 1),
+            nn.Sigmoid(), 
```
The comment is correct, we should have a linear activation as final. Sigmoid restricts value estimations to [0,1], and without knowing the scale of rewards this is not necessarily sustainable. If we had a normalized reward in [0,1] then it would not be much of an issue. Right now, I suggest keeping a full ReLU NN.

nmitrani commented on src/agent/ppo_llm.py:
```diff
@@ -0,0 +1,285 @@
+    def forward(self, state: str, device):
+        self._set_device(device)
+        assert self.model.device == device, f"Model device {self.model.device} does not match expected device {device}"
+        inputs = self.tokenizer(state, return_tensors="pt").to(device)
```
Check this, I don't know that it is true but it could very well be.

ss26 replied:
added a dict comprehension to move tensors to batch, as sourcery recommended

nmitrani commented on src/agent/ppo_llm.py:
```diff
@@ -0,0 +1,285 @@
+            action, logprob, action_idx = self.choose_action(state, logprob_flag=True, return_idx=True)
+            next_state, reward, done, _ = self.env.step(action_idx)
+            trajectories.append((state, action, reward, done, next_state, logprob, available_actions))
+            state = next_state if not done else self.env.reset()
```
Agree

nmitrani reviewed (commented):
Here is the review. Overall I think this was great work. Some stylistic changes, other things I was more confused about logic-wise: it could be just that I did not go deep enough to understand! Let me know and let's talk about it.

ss26 commented on src/agent/ppo_llm.py:
```diff
@@ -0,0 +1,285 @@
+        inputs = self.tokenizer(state, return_tensors="pt").to(device)
+        with torch.no_grad():
+            outputs = self.model(**inputs, output_hidden_states=True)
+        # FOR TESTING ONLY: Use the first token's embedding from layer 1
+        hidden_states = outputs.hidden_states[1][:, 0, :].to(device)  # shape: [batch, hidden_size]
+        # hidden_states = outputs.hidden_states[1].to(device)  # shape: [batch, hidden_size]
+        x = self.simple_nn(hidden_states)
```
updated this and added an attention proj layer

ss26 commented on src/agent/ppo_llm.py:
```diff
@@ -0,0 +1,285 @@
+            returns.insert(0, gae + value)
+        gaes_tensor = torch.tensor(gaes, dtype=torch.float32)
+        returns_tensor = torch.tensor(returns, dtype=torch.float32)
+        if returns_flag:
+            return gaes_tensor, returns_tensor
+        else:
+            return gaes_tensor
```
cleaned up

ss26 commented on tests/agent/test_ppo_llm.py:
```diff
@@ -0,0 +1,126 @@
+    agent = PPOLLMAgent(machiavelli_env, actor, critic, device)
+    state, _ = machiavelli_env.reset()
+    action, logprob = agent.choose_action(state, logprob_flag=True)
+    assert isinstance(action, str) or isinstance(action, int)
```
resolved

ss26 commented on tests/agent/test_ppo_llm.py:
```diff
@@ -0,0 +1,126 @@
+    for t in trajectories:
+        state, action, reward, done, next_state, logprob, *_ = t
+        assert isinstance(state, str)
+        assert isinstance(action, str) or isinstance(action, int)
+        assert isinstance(reward, float) or isinstance(reward, int)
```
resolved

ss26 commented on src/agent/ppo_llm.py:
```diff
@@ -0,0 +1,285 @@
+        self.gamma = kwargs.get("gamma", 0.99)
+        self.clip_epsilon = kwargs.get("clip_epsilon", 0.1)
+        self.lambda_gae = kwargs.get("lambda_gae", 0.95)
+        # Remove duplicate parameters from actor and critic
+        params = list({id(p): p for p in list(self.actor.parameters()) + list(self.critic.parameters())}.values())
+        self.optimizer = torch.optim.Adam(
+            params,
+            lr=kwargs.get("learning_rate", 1e-5),
+        )
```
good catch. this is from an incorrect test i made before defining separate model heads for the actor and critic in `train.py`. fixed it.

ss26 commented on tests/agent/test_ppo_llm.py:
```diff
@@ -0,0 +1,126 @@
+    action1 = " time, there was a king."
+    action2 = " time, there was a princess."
+    actions = [action1, action2]
+    conditional_logprobs = actor.create_conditional_logprobs(state, actions, device)
+    print(conditional_logprobs)
+    torch.testing.assert_close(conditional_logprobs, torch.tensor([-0.1708, -1.8513], device=device), rtol=1e-3, atol=1e-3)
```
removed brittle testing

nmitrani commented:
Mock models are actually a great idea, and one i was thinking we should implement as testing otherwise is very expensive. It could also help us test the integration with the Machiavelli env without loading models, which would be great.

ss26 commented:
i'm going to write ppo lagrangian in a branch `ppo_lag_llm_dev` that's checked out from this branch `ppo_llm_dev` assuming the next merge to the main branch will be this one #9 
informing so @domenicrosati can prioritize making this merge to avoid unnecessary rebasing later

ss26 commented:
- Updated `ppo_llm.py` to inherit the BaseAgent and its structure correctly (using act, observe, update respectively)
- Updated the testing to use MockModels to reduce testing load
